v-next/hardhat/src/internal/builtin-plugins/coverage/helpers.ts

```diff
@@ -0,0 +1,57 @@
+import type { HookContext } from "../../../types/hooks.js";
+import type { HardhatRuntimeEnvironmentImplementation } from "../../core/hre.js";
+
+import { getOrCreateGlobalHardhatRuntimeEnvironment } from "../../hre-initialization.js";
+
+/**
+ * This function unsafely casts HardhatRuntimeEnvironment to the internal
+ * HardhatRuntimeEnvironmentImplementation. We use it to access hre fields
+ * that we don't want to be exposed publicly through the HardhatRuntimeEnvironment
+ * interface.
+ *
+ * The use of this function should be limited to a minimum as it is inherently
+ * unsafe.
+ *
+ * @param context An instance of HookContext i.e. HardhatRuntimeEnvironment
+ * @returns A typed instance of HardhatRuntimeEnvironmentImplementation
+ */
+export function unsafelyCastAsHardhatRuntimeEnvironmentImplementation(
+  context: HookContext,
+): HardhatRuntimeEnvironmentImplementation {
+  const hre =
+    /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- we know this is the right implementation */
+    context as HardhatRuntimeEnvironmentImplementation;
+  return hre;
+}
+
+/**
+ * NOTE: The following helpers interact with the global HRE instance only;
+ * This is OK because:
+ * - They are intended for the internal use only. They are exposed via the
+ *   internal public API only.
+ * - We know the HRE has been initialized by the time they are used.
+ */
+
+export async function clearCoverageData(): Promise<void> {
+  const hre = await getOrCreateGlobalHardhatRuntimeEnvironment();
```

Comment:
I think doing `await import("../../.../index.js")` here would be better, as it ensure that the HRE has the same initialization as the one the user would expect. i.e. importing the same file that the user would import as `import hre from "hardhat"`

Reply:
Having reviewed the other PR now, maybe this is better done on a separate PR, to avoid the need for rebasing the other one.

Reply:
I merged the other PR into this one and then addressed this. I added the import and, instead of the cast, I did the assert of `HardhatRuntimeEnvironmentImplementation` instead. Similarly, to what you suggested with getting coverage manager from hook context.

```
const { default: hre } = await import("../../../index.js");
assertHardhatInvariant(hre instanceof HardhatRuntimeEnvironmentImplementation, "...");
```